Source/inv.cpp

```diff
@@ -347,50 +491,89 @@ void CheckInvPaste(Player &player, Point cursorPosition)
+				int8_t iv = std::abs(player.InvGrid[testCell]);
+				if (it != 0) {
+					if (it != iv) {
+						// Found two different items that would be displaced by the held item, can't paste the item here.
+						return std::nullopt;
 					}
 				} else {
-					it = -iv;
+					it = iv;
 				}
 			}
-		} else {
-			// check that the item we're pasting only overlaps one other item (or is going into empty space)
-			unsigned originCell = static_cast<unsigned>(slot - SLOTXY_INV_FIRST);
-			for (unsigned rowOffset = 0; rowOffset < static_cast<unsigned>(itemSize.height * InventorySizeInSlots.width); rowOffset += InventorySizeInSlots.width) {
-				for (unsigned columnOffset = 0; columnOffset < static_cast<unsigned>(itemSize.width); columnOffset++) {
-					unsigned testCell = originCell + rowOffset + columnOffset;
-					// FindTargetSlotUnderItemCursor returns the top left slot of the inventory region that fits the item, we can be confident this calculation is not going to read out of range.
-					assert(testCell < sizeof(player.InvGrid));
-					if (player.InvGrid[testCell] != 0) {
-						int8_t iv = std::abs(player.InvGrid[testCell]);
-						if (it != 0) {
-							if (it != iv) {
-								// Found two different items that would be displaced by the held item, can't paste the item here.
-								return;
-							}
-						} else {
-							it = iv;
-						}
-					}
+		}
+	}
+
+	return it;
+}
+
+std::optional<int8_t> GetPrevItemId(int slot, const Player &player, const Size &itemSize)
+{
+	int8_t item_something = 0;
+	int cell_num = slot - SLOTXY_INV_FIRST; // Where in InvGrid
+	if (player.HoldItem._itype == ItemType::Gold) {
+		// Placing a gold item into the grid somewhere???
+		if (player.InvGrid[cell_num] != 0) {
+			int8_t item_cell_begin = player.InvGrid[cell_num]; // item_cell_begin is not right name
+			if (item_cell_begin > 0) {
+				if (player.InvList[item_cell_begin - 1]._itype != ItemType::Gold) {
+					item_something = item_cell_begin;
 				}
+			} else {
+				item_something = -item_cell_begin;
 			}
 		}
-	} else if (il == ILOC_BELT) {
+	} else {
+		auto const maybe_overlapped_item = CheckOverlappingItems(slot, player, itemSize, item_something);
+		if (!maybe_overlapped_item) {
+			return std::nullopt;
+		}
+		item_something = *maybe_overlapped_item;
+	}
+
+	return item_something;
+}
+
+void CheckInvPaste(Player &player, Point cursorPosition)
+{
+	const Size itemSize = GetInventorySize(player.HoldItem);
+
+	const int slot = FindTargetSlotUnderItemCursor(cursorPosition, itemSize);
+	if (slot == NUM_XY_SLOTS)
+		return;
+
+	auto const desired_loc = player.GetItemLocation(player.HoldItem);
+	const auto il = GetItemEquipType(slot, desired_loc, player);
+	if (il == ILOC_BELT) {
 		if (!CanBePlacedOnBelt(player, player.HoldItem))
 			return;
-	} else if (desiredIl != il) {
+	}
+
+	if ((il != ILOC_UNEQUIPABLE) && (desired_loc != il)) {
+		return;
+	}
+
+	auto const maybe_it = (il == ILOC_UNEQUIPABLE) ? GetPrevItemId(slot, player, itemSize) : 0;

```

Comment:
Also a quite ambiguous variable name

Reply:
I agree, hence why this feedback request! I guess it gets the item ID from the inv grid? I was not too sure on this one.